module/icp/algs/sha2/sha2.c

```diff
@@ -53,17 +55,220 @@ static void Encode(uint8_t *, uint32_t *, size_t);
 static void Encode64(uint8_t *, uint64_t *, size_t);
 
 /* userspace only supports the generic version */
-#if	defined(__amd64) && defined(_KERNEL)
-#define	SHA512Transform(ctx, in) SHA512TransformBlocks((ctx), (in), 1)
-#define	SHA256Transform(ctx, in) SHA256TransformBlocks((ctx), (in), 1)
+#if	defined(_KERNEL)
+
+#include <sys/simd.h>
+
+typedef void (*sha256_block_f)(uint32_t *state, const void *in, size_t num);
+typedef void (*sha512_block_f)(uint64_t *state, const void *in, size_t num);
+
+#endif	/* _KERNEL */
+
+#if defined(_KERNEL)
+
+#define	BENCHMARK_BUFFER_SIZE (32 * 1024)
+
+// SHA-256
+
+static void sha256_alg_impl_benchmark(
+    const alg_impl_ops_t *ops, void *buffer, size_t buffer_n)
+{
+	SHA2_CTX sha2_ctx;
+	sha256_block_f sha256_impl = ops->ctx;
+	sha256_impl(sha2_ctx.state.s32, buffer, buffer_n >> 6);
+}
+
+static void
+SHA256TransformBlocksGeneric(uint32_t *state, const void *in, size_t num);
+
+static alg_impl_ops_t sha256_impl_generic = {
+	SHA256TransformBlocksGeneric, alg_impl_will_always_work, 0, "generic"};
+
+#if defined(__amd64)
+extern void
+sha256_x86_64_transform(uint32_t *state, const void *in, size_t num);
+
+static alg_impl_ops_t sha256_x86_64 = {
+	sha256_x86_64_transform, alg_impl_will_always_work, 1, "x86_64"};
+#endif
+
+#if defined(__amd64) && defined(HAVE_AVX)
+static boolean_t
+sha256_avx_will_work(void)
+{
+	return (kfpu_allowed() && zfs_avx_available());
+}
+
+extern void sha256_avx_transform(SHA2_CTX *ctx, const void *in, size_t num);
+static alg_impl_ops_t sha256_avx = {
+	sha256_avx_transform, sha256_avx_will_work, 10, "sha-avx"};
+#endif
+
+#if defined(__amd64) && defined(HAVE_SSSE3)
+static boolean_t
+sha256_ssse3_will_work(void)
+{
+	return (kfpu_allowed() && zfs_ssse3_available());
+}
+
+extern void sha256_ssse3_transform(SHA2_CTX *ctx, const void *in, size_t num);
+static alg_impl_ops_t sha256_ssse3 = {
+	sha256_ssse3_transform, sha256_ssse3_will_work, 30, "sha-ssse3"};
+#endif
+
+#if defined(__amd64) && defined(HAVE_SHA)
+static boolean_t
+sha256_ni_will_work(void)
+{
+	return (kfpu_allowed() && zfs_sha_available());
+}
+
+extern void sha256_ni_transform(SHA2_CTX *ctx, const void *in, size_t num);
+static alg_impl_ops_t sha256_ni = {
+	sha256_ni_transform, sha256_ni_will_work, 40, "sha-ni"};
+#endif
+
+/* All compiled in implementations */
+static const alg_impl_ops_t *sha256_all_impl[] = {
+	&sha256_impl_generic,
+#if defined(__amd64)
+	&sha256_x86_64,
+#endif
+#if defined(__amd64) && defined(HAVE_AVX)
+	&sha256_avx,
+#endif
+#if defined(__amd64) && defined(HAVE_SSSE3)
+	&sha256_ssse3,
+#endif
+#if defined(__amd64) && defined(HAVE_SHA)
+	&sha256_ni,
+#endif
+};
+
+static alg_impl_ops_t *sha256_supp_impl[ARRAY_SIZE(sha256_all_impl)];
```

Comment:
`static const alg_impl_ops_t *sha256_supp_impl`

Reply:
Fixed.

Reply:
Likewise

Reply:
This cannot be const, because `sha256_supp_impl` is filled in at runtime.